src/app.rs

```diff
@@ -144,16 +141,10 @@ impl Bound {
         // Wait for a signal to shutdown from explicit admin shutdown or signal
         self.shutdown.wait().await;
 
-        // Start a drain; this will wait for all drain_rx handles to be dropped before completing,
-        // allowing components to terminate.
-        // If they take too long, terminate anyways.
-        match time::timeout(self.config.termination_grace_period, self.drain_tx.drain()).await {
-            Ok(()) => info!("Shutdown completed gracefully"),
-            Err(_) => warn!(
-                "Graceful shutdown did not complete in {:?}, terminating now",
-                self.config.termination_grace_period
-            ),
-        }
+        // Start a drain; this will attempt to end all connections
+        // or itself be interrupted by a stronger TERM signal, whichever comes first.
+        self.drain_tx.drain().await;
+
```

Comment:
> This seems like it begins a "hard" drain immediately but I thought we were at least considering a solution with SIGTERM being a signal that begins starving the ztunnel of connections by trying to shift them.

The outcome of that discussion was that `hard` drain is the _only_ kind of drain _ztunnel_ can actually ever perform, with respect to the current connections it is aware of.

The other model was "refuse new connections for X period and do nothing with existing connections until the end of that period, at which point `hard` drain starts." - that's perfectly fine but that has nothing to do with ztunnel self-managing connection termination, that has to do with whatever is redirecting connections to ztunnel, and in that scenario the timeout would be configured outside of ztunnel.

Reply:
What about:

1. get SIGTERM & go into "terminating" status
2. ztunnel doesn't react, external control plane/CNI component/TBD notices status change and initiates traffic shifting if a new ztunnel on that node is up
3. once there are no connections (possibly + some slop time if we think necessary) or k8s deadline is reached pod goes away
4. Linux mops up the open connections (if there are any)

Was this dismissed? I think it provides a better outcome if someone performs some flavor of k8s terminate on an active ztunnel. I'm not sure relying on that kernel behavior is necessarily the best option but I didn't think we outright took it off the table.

Reply:
Downside would be short term while 2 is mostly wishful thinking the outcome might be worse and that idea of a 30h grace window for the pod to drain will just be problematic but its a step in that direction.

Reply:
> What about:
> 
>     1. get SIGTERM & go into "terminating" status
> 
>     2. ztunnel doesn't react, external control plane/CNI component/TBD notices status change and initiates traffic shifting if a new ztunnel on that node is up

This is what happens today with the Istio CNI - it proactively switches redirection to the newest ztunnel if there are more than one in READY state.

Additionally, if the CNI (or istiod, or whatever) itself is the thing that sends the SIGTERM, it can proactively manage this without having to notice.

>     3. once there are no connections (possibly + some slop time if we think necessary) or k8s deadline is reached pod goes away

Do you mean "no connections" as in old ztunnel is being starved by CNI (ours or native) redirecting new connections to the newer ztunnel instance, or "no connections" as in old ztunnel is actively RSTing it's active connections?

Former is something we can do later which would build on this, latter is what this PR does.

There isn't a workable solution that isn't one of these two variations. Node cordons can also work but wouldn't rely on monitoring ztunnel connections at all - they are more of a brute force option that relies on app pods being drained from the node.

>     4. Linux mops up the open connections (if there are any)
> 
We're relying on this no matter what if the platform deadline fires before ztunnel is "done" (whatever "done" means) - which ztunnel has no knowledge or control over.

> Was this dismissed? I think it provides a better outcome if someone performs some flavor of k8s terminate on an active ztunnel. I'm not sure relying on that kernel behavior is necessarily the best option but I didn't think we outright took it off the table.

Reply:
Sorry - I was misunderstanding - the suggestion here is to go slightly further and remove

```
   self.drain_tx.drain().await;
```

- that is, have ztunnel do _nothing_ on SIGTERM, other than exit when it gets to 0 connections, or await a more insistent followup signal.

That makes sense, that's probably a good followup, that also ultimately requires some more graceful connection handling than we currently have in CNI right now to be in place.

Reply:
Yeah. That's what I was asking about. Right now I think it's probably just about six of one half-dozen of the other but as this matures we may be able to provide a better experience if ztunnel doesn't RST immediately when it get's the SIGTERM. That way whatever process handles configuring ztunnel traffic capture can react to the `terminating` status on a ztunnel pod by starving it of connections and that starvation phase would have the duration of the k8s terminationGracePeriod.